[L1PF_10_6_X] Add dump and COE File Output

aperloff opened the pull request:
This PR is intended to re-add the ability to save binary (.dump) files which contain the discretized and regionized PF event information. It also adds the ability to save a COE file with just the track information.

In order to accomplish the task of adding the COE file with the 96-bit track word definition, the Xilinx ap_int and ap_fixed headers were also added to the PR. As far as I'm aware all of the licensing issues have been taken care of, but if someone has a comment or concern about this, now would be a good time to address this. On a more technical side, I put the files in the place that made the most logical sense to me. That doesn't mean it's the best/right place to put them. I'm open to suggestions. Also, in many places the Xilinx headers don't follow the CMS coding rules and therefore generate many warnings. I think we will have to edit the files to conform with the CMS rules before this could be merged into the official CMSSW repo.

Commit:
This commit adds back the ability to save a regionized binary (.dump) file for storing events. It also adds the new ability to store a track only text-based (.coe) file. There is a new executable which will compare the ROOT, dump, and COE files for consistency. This new executable can be run with 'scram b runtests'.

This is a combination of 10 commits that are being squashed into a single commit. Not only is this cleaner, but some files were committed which are not being removed from the history. In order to preserve the more granular commit history the 10 messages below show the individual commit messages:

- Add the Xilinx arbitrary precision int and float types from Vivado 2019.1.
- Add back the ability to save a regionized binary (.dump) file for storing events as well as the new ability to store a track only text-based (.coe) file.
- Add a new executable for testing the various output formats for PF. It will compare the ROOT, dump, and COE files. It can be run with 'scram b runtests'.
- Slightly modify the configuration of the regions and add in a track standalone (TSA) region.
- Rename the code that tests the output files from the L1TPFProducer.
- Return the spacing back to it's original state before the most recent PR.
- Remove the l1ParticleFlow_pf_tsa module from the default sequence.
- Slightly modify the compare_lv_with_tolerance function signature.
- Remove all references to ap_int and instead use boost and std objects/functions for converting between std ints and binary representations. Also move the COE file handling to a separate class.
- Remove the ap_int, ap_fixed, and associated headers. These will be included in CMSSW some other way.

gpetruc commented on L1Trigger/Phase2L1ParticleFlow/python/l1ParticleFlow_cff.py:
```diff
@@ -219,6 +252,7 @@
     l1ParticleFlow_pf_barrel +
     l1ParticleFlow_pf_hgcal +
     l1ParticleFlow_pf_hf +
+    l1ParticleFlow_pf_tsa +
```
I think if this instance is used only to make dump files for the tracker-only workflow it shouldn't be run in the standard sequence (it's good to keep it in the cff, just not in the sequence(

aperloff replied:
I have taken it out of the sequence and moved it into my own, private configuration file. That way it will not show up in the sequence for the purposes of the PR.

gpetruc commented on L1Trigger/Phase2L1ParticleFlow/plugins/L1TPFProducer.cc:
```diff
@@ -22,254 +24,375 @@
+					current_track = regions[iphi].track[irow];
+					current_track_word = (ap_uint<14>(0), ap_uint<7>(current_track.hwStubs), ap_uint<3>(0), ap_uint<4>(current_track.hwChi2),
+										  ap_int<13>(0), ap_int<12>(current_track.hwZ0), ap_int<16>(current_track.hwVtxEta), ap_int<12>(current_track.hwVtxPhi),
+										  current_track.hwCharge,ap_uint<14>(current_track.hwPt));
```
in the raw track word, it's it stored as 1/pt ?

gpetruc replied:
also, I had understood the raw track word was now accessible from the track object itself.

aperloff replied:
The updated TTTrack, and hence the TTTrackWord, is not yet available. When that happens, I will update this code to use the official track word. When that happens, the pT will be updated to be q/R (the actual quantity stored) and eta will be updated to be tan(lambda). I don't currently have access to R or to lambda, so I chose to keep pT and eta for now.

gpetruc commented on L1Trigger/Phase2L1ParticleFlow/test/testOutputFiles.cpp:
```diff
@@ -0,0 +1,452 @@
+	return file;
+}
+
+bool compare_lv_with_tolerance(TLorentzVector a, TLorentzVector b, std::vector<float> tolerance = {0,0,0}) {
```
can't it be const std::vector<float> & ?
also, here it's initialized with 3 values but below you use tolerance[4]

aperloff replied:
You are right. I will fix this in my next commit.

gpetruc reviewed (commented):
Hi,

If the ap_int are only used to stack bits together when writing the file, I would prefer if that could be implemented using the std or boost libraries, as I would prefer to wait to add the ap_int headers in the branch for the moment (there should be some more general discussion of where they should go, I think they have to be added as an external, we can't just copy them in the CMSSW source tree since the licensing is different)

There's a few other general items:
 * Don't re-indent the code in the PFProducer: it unecessarily complicates diffs, rebases, etc.
 * For the COE file handling, I would prefer if it could be factored out in another class to reduce the amount of code that is copied in the PF producer

More comments inline.

I have to update the 10_6_X branch to the new upstream one from Vladimir, so maybe this can be re-applied on the updated branch?

Giovanni

aperloff commented:
Hi @gpetruc, I think I have addressed all of your previous comments.
1) The spacing should be back to the state it was in before
2) The Xilinx files have been removed from the PR (I removed the files and squashed the commits to remove them from the history). I have also removed all references to them and replaced the functionality with a combination of std and boos libraries.
3) The COE file handling has been factored out into a new COEFIle class.
4) I addressed all of your inline comments.
I can re-applying this on an updated 10_6_X, but do you know when that might be?

gpetruc commented:
Thanks!

Will have a look tomorrow (I'm still in the process of getting a working
branch on the new 10_6_X)

Giovanni

Il Gio 25 Lug 2019, 20:30 Alexx Perloff <notifications@github.com> ha
scritto:

> @aperloff <https://github.com/aperloff> requested your review on: #17
> <https://github.com/p2l1pfp/cmssw/pull/17> [L1PF_10_6_X] Add dump and COE
> File Output.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/p2l1pfp/cmssw/pull/17?email_source=notifications&email_token=ABDNWR574MXIAAIBBTLJW2DQBHWMPA5CNFSM4IFHMOCKYY3PNVWWK3TUL52HS4DFWZEXG43VMVCXMZLOORHG65DJMZUWGYLUNFXW5KTDN5WW2ZLOORPWSZGOSWRFSUI#event-2510444881>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABDNWR2CWCSUKZZPAUMEPW3QBHWMPANCNFSM4IFHMOCA>
> .
>

gpetruc reviewed (approved):
Code looks ok to me now, thanks!

Commit:
Improve the integration test code to use the l1tpf_impl::Regions rather than it's own sector format. This follows the pattern of the RegionMapper and allows the code to have multiple regions in eta as well as phi.
